### DISCUSSION TITLE
Unicode/Latin2/ISO_8859_2 characters like É Á Ő Ű etc

### DISCUSSION THREAD

**Discussion by nagyrobi** [HASwitchPlate/openHASP#74] [Q&A] [answered]
Trying this example on Lanbon L8: https://fvanroie.github.io/hasp-docs/#example-2/
Noticed that special characters in word "Átvíztűrőtükörfúrógép" are skipped if I send them over via MQTT.
I found the [description](https://fvanroie.github.io/hasp-docs/#fonts/) about built-in fonts bit that only specifies how to set the size, not the font itself. Any hint on how to select it correctly? The dropdown box is empty in the web interface.

**Comment by nagyrobi** [comment]
Uploading `HMI UbuntuCondensed-Regular 12 (ISO_8859_2).zi` from your other repository unsuccessful.
Renamed file to something without spaces and that resulted in a bootloop after reboot, only `NETWIZE` logo is flashing. :(

EDIT: I'll try with recompiling.

**Comment by nagyrobi** [comment]
It looks like `robotocondensed_regular_XX_nokern.c` is completely missing the `0x0A0-0x0FF` Latin-1 Supplement and `0x100-0x17F` Latin Extended-A characters.

But in all sizes it includes the FontAwesome5 symbols.

:(

**Reply by nagyrobi** [reply]
Font handling could be optimized a bit:
- pre-create `.c` arrays with various font selections and sizes and make them selectable through `user_config_override.h` or `platformio-override.ini`
- don't include FontAwesome in the same files with the normal characters and in all sizes, instead, create separate FontAwesome font files of multiple sizes. Typically on a screen, icons look nice if they are all the same size, including them in multiple sizes is a waste of space imho. One can still recompile with different icon sizes if they want to. Or perhaps leave out FontAwesome completely and offer it as uploadable .zi files. Maybe someone would like MDI instead...
- also including all characters of all languages can be space-consuming so maybe it should be considered creating font files containing only the characters specific to languages. Like translations are done. One could select the preferred language font files through `user_config_override.h` or `platformio-override.ini`. Ask for contributions from users with suggested language files with their characters. I can contribute two.
- have a separate, optional font file containing only numbers, dots etc., for displaying numeric values [bigger](https://github.com/fvanroie/hasp-lvgl/discussions/37#discussioncomment-439665) (like size 34 or 44). Suggestion: `0x020,0x021,0x02B-0x02E,0x030-0x03A,0x03F`.

Based on the above I was able to add support for my language's diacritics, include an extrabig number set and free up 121,8kB of flash space.

**Reply by fvanroie** [reply]
Much like LVGL, the icons and characters are _purposefully_ put into the same font file to allow you to mix icons and text seamlessly. Also, the character set is the same in all font sizes to allow you to easily scale up or down by just changing the point size. No need to know which letter is in which font, either the character is supported or it is not...

For the default built-in fonts, simplicity is paramount. Having icons/characters dependent on font and size will add confusing and complexity. I can imagine the default character set to be dependent on the firmware language. When you install the HU version "Átvíztűrőtükörfúrógép" are also available in the default built-in fonts, which seems reasonable and manageable.

Some middle ground will have to be found between added characters, icons and font sizes without overly complicating the number of available firmware options.

As always, users are free to compile their own customized versions of the firmware including the fonts.
The default firmware is a general purpose solution, that should fit *most* use-cases.

**Reply by nagyrobi** [reply]
Agree, but how will this all be possible when we already have 95% of the flash full? My approach was to try to save as much space as possible because we'ver ran into issues going over 100%...

**Reply by fvanroie** [reply]
Flash is not the main issue, we can re-partition it to increase the firmware size, while reducing the user portion of the filesystem. The constraint is RAM because **all** characters are loaded into RAM at boot whether you use them or not...

As with all micro-controllers, you need to balance features and performance with the resources available.
Trying to find an acceptable _least common denominator_ is what is the difficult part.

**Comment by nagyrobi** [comment]
Here is my suggestion for a basic character set, which removes usually unused (at least less likely to be used on hasp-lvgl displays) characters from Basic Latin-1, and adds some useful from Latin-1 Supplement (like degree symbol, +/- sign, micro, quarters:
`0x020,0x021,0x025,0x028-0x03A,0x03C-0x05B,0x05D,0x05F,0x061-0x07A,0x07C,0x0B0,0x0B1,0x0B5,0x0B7,0x0BC-0x0BE`

New font files in `.c` array format can be generated with an online tool: https://lvgl.io/tools/fontconverter
Tips:
- use your desired font in `ttf` or `woff` format
- Name: type with all small caps the desired name and size (e.g. `roboto_condensed_12`)
- Size: 12, Bpp: 2 bit-per-pixel
- Range: a comma-separated list of the characters desired in the font file, like above. Character codes [list](https://jrgraphix.net/r/Unicode/) (note the code numbers show without `0x`)

**Comment by nagyrobi** [comment, marked as answer]
Submitted a PR with fomt-selector from `user_config_override.h`: https://github.com/fvanroie/hasp-lvgl/pull/83

**Comment by fvanroie** [comment]
While you can certainly do this to fit your individual needs, the goal of the default firmware is to get a device set up easily and usable.
Some users want more icons and ascii-only, while other users might want Cyrillic characters and no icons...

Additional fonts are on the roadmap, to be added on-the-fly to the filesystem instead of compiling in the firmware.
.zi fonts are currently broken, and this needs to be investigated and fixed. Lvgl also has a way to dynamically load fonts that can help fix this problem.

Changing fonts using #defines seems hard to manage going forward, there are so many different language scripts.
I'm aiming to have a default _minimal_ font that supports the basic setup and examples, with configurable extra fonts on the filesystem.

**Reply by nagyrobi** [reply]
I have thought so, that's why it's implemented so that the original methods aren't affected. I'd keep this until .zi fonts get a fix, and until dynamic loading gets implemented...

I didn't know that it's known that .zi are broken per se, all I've noticed that uploading and setting a zi font creates a bootloop which can be recovered by flashing only. I thought maybe I'm doing somehing wrong.

**Comment by fvanroie** [comment]
I have found this [overview of alphabets](https://en.wiktionary.org/wiki/Appendix:Latin_script/alphabets#French_alphabet). It might be a useful exercise to determine which letters are common in multiple alphabets and languages. I can run a script to generate the .c files for different fonts and compare the file sizes.

Common letters can be added to the default font, but some icons probably need to go... it's a balancing act, but I'm open for suggestions.

**Reply by nagyrobi** [reply]
My suggestion [above](https://github.com/fvanroie/hasp-lvgl/discussions/74#discussioncomment-439162) does exactly that. Even the basic latin set contains useless characters for a display, I took instead degree symbol, for example. That's a good start for a minimal approach that you wish.

For Hungarian, only these need to be added:
`0x0C1,0x0C9,0x0CD,0x0D3,0x0D6,0x0DA,0x0DC,0x0E1,0x0E9,0x0ED,0x0F3,0x0F6,0x0FA,0x0FC,0x150,0x151,0x170,0x171`
For Romanian these:
`0x0C2,0x0CE,0x0E2,0x0EE,0x102,0x103,0x15E,0x15F,0x162,0x163,0x218,0x219,0x21A,0x21B`
There's no overlapping between the two...

You could check your generated .c files with the ones I've submitted in the PR.

**Reply by fvanroie** [reply]
I *might* see this working maybe in combination with the [localization of the debug messages](https://github.com/fvanroie/hasp-lvgl/tree/0.4.0-dev/src/lang).
There is an idea to have a local version for HU, RO, ... and those firmware could then include the **additional** characters for that language.

For each language, we then need to maintain a list of extra glyphs to add into the font.

Icons should be the same across all firmwares because they are universal and don't depend on the language spoken.

**Reply by nagyrobi** [reply]
Will.submit localized versions of these for Hu, Ro.

**Reply by nagyrobi** [reply]
I see both EN and NL languages start with

```
#ifndef HASP_LANG_EN_US_H
#define HASP_LANG_EN_US_H
```
Is that on purpose?

**Reply by nagyrobi** [reply]
PRs submitted

**Reply by nagyrobi** [reply]
If you change in `lang.h` to `#include "hu_HU.h"`
![kép](https://user-images.githubusercontent.com/1550668/110615731-042af880-8194-11eb-92ea-2bc5b48089a6.png)

Also the WiFi setup first screen loos nice.

**Reply by nagyrobi** [reply]
I see lots of strings on the web interface which could be translated but are not (yet) in the language file. I think it's not a problem to maintain new strings appearing.

**Reply by fvanroie** [reply]
> Will.submit localized versions of these for Hu, Ro.

Thanks! Those files have been merged now.

> I see lots of strings which could be translated but are not (yet) in the language file

Indeed, this is a work in progress and I add new ones directly and old ones are moved when I encounter them

**Reply by nagyrobi** [reply]
Would you like me to create a configurable option for choosing the language in `user_config_override.h`? That would automatically select the correct font files containing the appropriate characters and language files. On the same principle as my previous PR for selecting the fonts.

**Comment by fvanroie** [comment]
I generate the fonts with a script and I'm adding the ranges to it to see what is the difference in file size.

**Reply by fvanroie** [reply]
Adding most latin international characters to the base font will add about 120+ characters and increase the font size by 50%. Consequently a lot of custom icons would have to be removed again. A quick ask on the [Discord channel](https://discord.gg/VCWyuhF) shows that not all users want localized characters and some prefer icons over text for touch displays... so there is a range of opinions.

So, the idea is now to make the default character set dependent on the firmware language installed.

**Reply by nagyrobi** [reply]
This is great.

**Comment by fvanroie** [comment]
I'm thinking of compiling a list like [this one](https://fvanroie.github.io/hasp-docs/#characters/) for the characters used in different languages. The CSV can be easily opened in a spreadsheet to be edited or filtered.

Then that list can be used to generate the fonts automatically e.g. font/hu/robotocondensed_regular_12_nokern.c
If possible, I want to keep the variable names of the fonts equal accross all languages, so lv_conf_v7.h doesn't need a lot of #ifdef's, but would just refer to `robotocondensed_regular_xx_nokern` regardless of the language selected.

This will need some clever src_filter or keep the files in a separante repo to include in a custom environment... not sure yet :)

**Reply by nagyrobi** [reply]
I suggest an extra column before the first language, called something like "present in default" or "present in all". To be able to show which symbols should be present or not, regardless of the language. 

Or should we mark these in the same columns as the language? For example every language should have the non-breaking space, and the degree symbol...

**Reply by fvanroie** [reply]
I suggest to just add all languages to that line. That is easiest way for a script to read the file without having to check another column...

**Reply by nagyrobi** [reply]
And the standard Latin-1 characters like ABCDEF...?

**Reply by nagyrobi** [reply]
> I suggest to just add all languages to that line. That is easiest way for a script to read the file without having to check another column...

PR done

**Reply by fvanroie** [reply]
Ascii 0x20 - 0x7E will be added automatically as these are also used by LVGL.
The list is for the extended characters on top of ASCII.

**Comment by fvanroie** [comment]
This is tied to the new material design icons and those new fonts will be available with other character-sets as well.
